File: 7. Wave Traversal.cpp
// Topic: Wave Traversal (column-wise up-down)
#include <bits/stdc++.h>
using namespace std;
int main()
{
    vector<vector<int>> A = {{1, 2, 3}, {4, 5, 6}, {7, 8, 9}};  // 3x3
    int n = A.size(), m = A[0].size();
    for (int j = 0; j < m; ++j)
    {
        if (j % 2 == 0)
            for (int i = 0; i < n; ++i) cout << A[i][j] << " ";
        else
            for (int i = n - 1; i >= 0; --i) cout << A[i][j] << " ";
    }
    cout << "\n";
}

// TC: O(n*m)
// SC: O(1)

File: 7. Wave Traversal.py
"""Topic: Wave Traversal"""
def wave(A):
    n=len(A); m=len(A[0])
    out=[]
    for j in range(m):
        if j%2==0:
            for i in range(n): out.append(A[i][j])
        else:
            for i in range(n-1,-1,-1): out.append(A[i][j])
    return out

if __name__=='__main__':
    A=[[1,2,3],[4,5,6],[7,8,9]]
    print(*wave(A))

# TC: O(n*m)
# SC: O(n*m)
